File: csharp-dotnet/tests/Architecture.Shell.Cqrs.Tests/MediatorTests.cs
using Architecture.Core.Functional;
using Architecture.Shell.Cqrs;
using Moq;

namespace Architecture.Shell.Cqrs.Tests;

/// <summary>
/// UT-001: Handler Registration Uniqueness Tests
/// UT-001c: Handler Uniqueness at Constructor Time Tests
/// </summary>
public class MediatorTests
{
    // UT-001: T026
    [Fact]
    public void Should_ThrowException_When_ZeroHandlersRegistered()
    {
        // Given: No handlers registered for a specific command type
        var serviceProvider = new Mock<IServiceProvider>();
        var logger = new Mock<Microsoft.Extensions.Logging.ILogger<IMediator>>();

        // Configure service provider to return empty handler collection
        serviceProvider
            .Setup(sp => sp.GetService(It.IsAny<Type>()))
            .Returns((object?)null);

        // When: Mediator is created
        var exception = Record.Exception(() =>
        {
            var mediator = new Mediator(serviceProvider.Object, logger.Object);
            // Attempt to send a command that has no handler
            var command = new TestCommandWithNoHandler();
            var task = mediator.SendAsync(command, CancellationToken.None);
            task.Wait();
        });

        // Then: Should throw an exception indicating no handler found
        exception.Should().NotBeNull();
    }

    // UT-001: T031
    [Fact]
    public void Should_ThrowException_When_MultipleHandlersRegistered()
    {
        // Given: Multiple handlers registered for same command type
        var serviceProvider = new Mock<IServiceProvider>();
        var logger = new Mock<Microsoft.Extensions.Logging.ILogger<IMediator>>();

        // Configure to return multiple handlers (ambiguous)
        var handler1 = new Mock<ICommandHandler<MediatorTestCommand>>();
        var handler2 = new Mock<ICommandHandler<MediatorTestCommand>>();
        var handlers = new object[] { handler1.Object, handler2.Object };

        serviceProvider
            .Setup(sp => sp.GetService(typeof(IEnumerable<ICommandHandler<MediatorTestCommand>>)))
            .Returns(handlers);

        // When: Attempting to create mediator or send command
        var exception = Record.Exception(() =>
        {
            var mediator = new Mediator(serviceProvider.Object, logger.Object);
            var command = new MediatorTestCommand();
            var task = mediator.SendAsync(command, CancellationToken.None);
            task.Wait();
        });

        // Then: Should throw exception about ambiguous handler registration
        exception.Should().NotBeNull();
    }

    // UT-001: T036
    [Fact]
    public async Task Should_ResolveHandler_When_ExactlyOneHandlerRegistered()
    {
        // Given: Exactly one handler registered for command type
        var serviceProvider = new Mock<IServiceProvider>();
        var logger = new Mock<Microsoft.Extensions.Logging.ILogger<IMediator>>();
        var handler = new Mock<ICommandHandler<MediatorTestCommand>>();

        handler
            .Setup(h => h.HandleAsync(It.IsAny<MediatorTestCommand>(), It.IsAny<CancellationToken>()))
            .ReturnsAsync(Result.Ok());

        serviceProvider
            .Setup(sp => sp.GetService(typeof(IRequestHandler<MediatorTestCommand, Result>)))
            .Returns(handler.Object);

        var mediator = new Mediator(serviceProvider.Object, logger.Object);
        var command = new MediatorTestCommand();

        // When: Command is sent through mediator
        var result = await mediator.SendAsync(command, CancellationToken.None);

        // Then: Handler should be invoked successfully
        result.Should().NotBeNull();
        handler.Verify(h => h.HandleAsync(command, It.IsAny<CancellationToken>()), Times.Once);
    }

    // UT-001c: T036b
    [Fact]
    public void Should_ThrowException_When_ConstructorDetectsAmbiguousHandlers()
    {
        // Given: Service provider configured with 2+ handlers for same command type
        var serviceProvider = new Mock<IServiceProvider>();
        var logger = new Mock<Microsoft.Extensions.Logging.ILogger<IMediator>>();

        var handler1 = new TestCommandHandler1();
        var handler2 = new TestCommandHandler2();
        var handlers = new ICommandHandler<MediatorTestCommand>[] { handler1, handler2 };

        serviceProvider
            .Setup(sp => sp.GetService(typeof(IEnumerable<ICommandHandler<MediatorTestCommand>>)))
            .Returns(handlers);

        // When: Mediator constructor/build is called with validator that detects ambiguous handlers
        var exception = Record.Exception(() =>
        {
            var mediator = new Mediator(serviceProvider.Object, logger.Object, sp =>
            {
                // Validator checks for ambiguous handler registration
                var commandHandlers = sp.GetService(typeof(IEnumerable<ICommandHandler<MediatorTestCommand>>)) as IEnumerable<ICommandHandler<MediatorTestCommand>>;
                if (commandHandlers != null && commandHandlers.Count() > 1)
                {
                    var handlerNames = string.Join(", ", commandHandlers.Select(h => h.GetType().Name));
                    throw new InvalidOperationException(
                        $"Ambiguous handler registration detected for 'TestCommand'. " +
                        $"Found multiple handlers: {handlerNames}");
                }
            });
        });

        // Then: Should throw with handler names in error message
        exception.Should().NotBeNull();
        exception.Should().BeOfType<InvalidOperationException>();
        exception.Message.Should().Contain(nameof(TestCommandHandler1));
        exception.Message.Should().Contain(nameof(TestCommandHandler2));
    }
}

// Test domain types
public record MediatorTestCommand : ICommand;
public record TestCommandWithNoHandler : ICommand;

public class TestCommandHandler1 : ICommandHandler<MediatorTestCommand>
{
    public Task<Result> HandleAsync(MediatorTestCommand request, CancellationToken cancellationToken)
    {
        return Task.FromResult(Result.Ok());
    }
}

public class TestCommandHandler2 : ICommandHandler<MediatorTestCommand>
{
    public Task<Result> HandleAsync(MediatorTestCommand request, CancellationToken cancellationToken)
    {
        return Task.FromResult(Result.Ok());
    }
}

File: java-spring/architecture-shell-cqrs/src/test/java/com/architecture/shell/cqrs/MediatorTests.java
package com.architecture.shell.cqrs;

import com.architecture.core.functional.Result;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Unit tests for Mediator handler registration and resolution.
 * Tests follow TDD approach - written before implementation exists.
 * Naming: Should_ExpectedBehavior_When_StateUnderTest
 */
class MediatorTests {

    /**
     * T027: UT-001 - Handler Registration Uniqueness
     * Given: No handlers registered for a command type
     * When: Mediator attempts to send command with no handler
     * Then: Should throw exception indicating zero handlers registered
     */
    @Test
    void should_ThrowException_When_ZeroHandlersRegistered() {
        // Given: Empty handler registry
        var handlerRegistry = new HandlerRegistry();
        var mediator = new MediatorImpl(handlerRegistry);

        // When: Attempting to send command with no registered handler
        // Then: Should throw IllegalStateException at runtime
        assertThatThrownBy(() -> mediator.send(new TestCommand()))
            .isInstanceOf(IllegalStateException.class)
            .hasMessageContaining("No handler registered")
            .hasMessageContaining("TestCommand");
    }

    /**
     * T032: UT-001 - Handler Registration Uniqueness (Multiple Handlers)
     * Given: Multiple handlers registered for same command type
     * When: Mediator is constructed
     * Then: Should throw exception indicating ambiguous registration
     */
    @Test
    void should_ThrowException_When_MultipleHandlersRegistered() {
        // Given: Two handlers for the same command type
        var handlerRegistry = new HandlerRegistry();
        handlerRegistry.register(new TestCommandHandler1());
        handlerRegistry.register(new TestCommandHandler2());

        // When/Then: Mediator construction should fail with ambiguous handler error
        assertThatThrownBy(() -> new MediatorImpl(handlerRegistry))
            .isInstanceOf(IllegalStateException.class)
            .hasMessageContaining("Multiple handlers registered")
            .hasMessageContaining("TestCommand")
            .hasMessageContaining("TestCommandHandler1")
            .hasMessageContaining("TestCommandHandler2");
    }

    /**
     * T037b: UT-001c - Handler Uniqueness at Constructor Time
     * Given: Multiple handlers registered for same command type
     * When: Mediator constructor/build is called
     * Then: Should throw exception with handler names BEFORE runtime
     */
    @Test
    void should_ThrowException_When_ConstructorDetectsAmbiguousHandlers() {
        // Given: Registry with 2 handlers for the same command type
        var handlerRegistry = new HandlerRegistry();
        var handler1 = new TestCommandHandler1();
        var handler2 = new TestCommandHandler2();
        handlerRegistry.register(handler1);
        handlerRegistry.register(handler2);

        // When: Mediator is constructed
        // Then: Should throw at construction time (not runtime) with handler names
        assertThatThrownBy(() -> new MediatorImpl(handlerRegistry))
            .isInstanceOf(IllegalStateException.class)
            .hasMessageContaining("Multiple handlers")
            .hasMessageContaining("TestCommandHandler1")
            .hasMessageContaining("TestCommandHandler2");
    }

    // Test fixtures (will fail compilation until interfaces are created)

    /**
     * Test command for handler registration tests
     */
    private static class TestCommand implements Command<Result<Void>> {
    }

    /**
     * First test handler
     */
    private static class TestCommandHandler1 implements CommandHandler<TestCommand, Result<Void>> {
        @Override
        public Result<Void> handle(TestCommand command) {
            return Result.success();
        }
    }

    /**
     * Second test handler (creates ambiguity)
     */
    private static class TestCommandHandler2 implements CommandHandler<TestCommand, Result<Void>> {
        @Override
        public Result<Void> handle(TestCommand command) {
            return Result.success();
        }
    }
}
